vm/src/org/mozartoz/truffle/nodes/call/CallMethodNode.java

```diff
@@ -43,89 +46,130 @@ public static CallMethodNode create() {
+			return MethodLookupNodeGen.create(null, null);
+		}
 
-		final Object name = labelNode.executeLabel(message);
-		assert OzGuards.isLiteral(name);
+		public abstract Lookup executeLookup(OzObject self, Object name);
+
+		@Specialization(guards = {
+				"self.getClazz() == cachedClazz",
+				"name == cachedName",
+		}, limit = "methodCache(3)")
+		protected Lookup descriptorIdentityLookup(OzObject self, Object name,
+				@Cached("self.getClazz()") DynamicObject cachedClazz,
+				@Cached("name") Object cachedName,
+				@Cached("genericLookup(self, name)") Lookup cachedLookup) {
+			return cachedLookup;
+		}
+
+		BranchProfile otherwiseProfile = BranchProfile.create();
+
+		@Specialization(contains = "descriptorIdentityLookup")
+		protected Lookup genericLookup(OzObject self, Object name) {
+			OzProc proc = getMethod(self, name);
+			if (proc == null) {
+				otherwiseProfile.enter();
+				proc = getMethod(self, OTHERWISE);
+				if (proc == null) {
+					CompilerDirectives.transferToInterpreterAndInvalidate();
+					throw new RuntimeException(name + " not found");
+				}
+				return new Lookup(proc, true);
+			}
+			return new Lookup(proc, false);
+		}
 
-		OzProc method = getMethod(self, name);
-		if (otherwise.profile(method == null)) { // redirect to otherwise
-			method = getMethod(self, "otherwise");
-			message = OTHERWISE_MESSAGE_FACTORY.newRecord(message);
+		@TruffleBoundary
+		protected static OzProc getMethod(OzObject self, Object name) {
+			DynamicObject methods = (DynamicObject) self.getClazz().get(ooMeth);
+			return (OzProc) methods.get(name);
 		}
 
-		Object[] arguments = new Object[] { self, message };
-		return callNode.call(frame, method.callTarget, OzArguments.pack(method.declarationFrame, arguments));
 	}
 
-	static final OzUniqueName ooMeth = OzUniqueName.get("ooMeth");
+	public static class Lookup {
+		public final OzProc proc;
+		public final boolean embed;
 
-	@TruffleBoundary
-	protected OzProc getMethod(OzObject self, Object name) {
-		DynamicObject methods = (DynamicObject) self.getClazz().get(ooMeth);
-		return (OzProc) methods.get(name);
-	}
+		private Lookup(OzProc proc, boolean embed) {
+			this.proc = proc;
+			this.embed = embed;
+		}
 
-	protected Object labelFromArgs(Object[] args) {
-		return labelNode.executeLabel(derefNode.executeDeref(args[0]));
 	}
 
-	protected OzProc methodFromArgs(OzObject self, Object[] args) {
-		return getMethod(self, labelFromArgs(args));
-	}
+	@ImportStatic(CallMethodNode.class)
+	@NodeChildren({ @NodeChild("self"), @NodeChild("lookup"), @NodeChild("args") })
+	public abstract static class MethodDispatchNode extends OzNode {
+
+		public static MethodDispatchNode create() {
+			return MethodDispatchNodeGen.create(null, null, null);
+		}
+
+		public abstract Object executeDispatch(VirtualFrame frame, OzObject self, Lookup lookup, Object[] args);
+
+		@Specialization(guards = {
+				"lookup.proc == cachedLookup.proc",
+				"!lookup.embed",
```

Comment:
This extra check sounds sub-optimal.
I would try to model this as just passing the OzProc, and if it's null, create a new lookup node for otherwise and a dispatch node and do the otherwise call in this node.